### PR TITLE
Add btrfs driver

### PR DESCRIPTION
**Pull request by alexlarsson**
Ok, so i should be doing that presentation.... but its so boring...
Procastrination is a great way to get shit done...

Here is a btrfs backend for the new plugin system. Seems to work fine on my laptop.

### CODE REVIEW

File: graphdriver/btrfs/btrfs.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+	var buf syscall.Statfs_t
+	if err := syscall.Statfs(rootdir, &buf); err != nil {
+		return nil, err
+	}
+
+	if buf.Type != 0x9123683E {
+		return nil, fmt.Errorf("%s is not a btrfs filesystem", rootdir)
+	}
+
+	d := &Driver{
```
Comment (creack) [inline graphdriver/btrfs/btrfs.go]: Why using a variable for the return?

File: graphdriver/btrfs/btrfs.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	"syscall"
+	"unsafe"
+)
+
+func init() {
+	graphdriver.Register("btrfs", Init)
+}
+
+func Init(home string) (graphdriver.Driver, error) {
+	rootdir := path.Dir(home)
```
Comment (creack) [inline graphdriver/btrfs/btrfs.go:30]: Those two lines could be grouped in a var() statement

File: graphdriver/driver.go [hunk 3]
```diff
@@ -39,6 +39,7 @@ var (
 	// Slice of drivers that should be used in an order
 	priority = []string{
 		"aufs",
+		"btrfs",
```
Comment (crosbymichael) [inline graphdriver/driver.go]: I'm guessing we should move this below devicemapper ?

Comment (alexlarsson) [inline graphdriver/driver.go]: For now i think we should have it below dummy so that its only enabled when you manually specify it.

Comment (shykes) [inline graphdriver/driver.go]: yes let's move it to the bottom.

**Comment by shykes** [comment]
I think we can ship this with 0.7.0, it's pretty easy to keep it neutral, and makes it more obvious how easy it is to add drivers.

TODOs:
- [ ] move to the bottom of driver list
- [ ] mention as experimental in the comments
- [ ] test that it doesn't break the rest

**Comment by alexlarsson** [comment]
Any motion on this?

**Comment by creack** [comment]
I'll address this after 0.7.1

**Comment by tianon** [comment]
ping - we've got lots of people curious about this (and it probably needs to move to the main repo now)

**Comment by crosbymichael** [comment]
@alexlarsson Can you move this PR to the main repo?  Whenever you have time

**Comment by alexlarsson** [comment]
Moved the PR to the main repo at: 
https://github.com/dotcloud/docker/pull/3718

Closing this PR.

**Pull request closed without merging**